### PR TITLE
Run periodic jobs against github.com/moby/moby

### PR DESCRIPTION
**Pull request by AshwinHIBM**
Docker no longer runs periodic test jobs on moby/moby for ppc64le. Instead, we plan to run these on our production cluster.
There are 4 jobs:
1. Check configuration
2. Build the dev image
3. Run unit tests
4. Run integration tests
Please refer https://github.com/ppc64le-cloud/docker-ce-build/pull/182 for scripts.

### CODE REVIEW

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        - base_ref: main
+          org: ppc64le-cloud
+          repo: docker-ce-build
+      reporter_config:
+          slack:
+            channel: 'prow-job-notifications'
+            report_template: 'Job {{.Spec.Job}} of type {{.Spec.Type}} ended with state {{.Status.State}}. <!subteam^S02N6DWBX0F> <{{.Status.URL}}|View logs>'
+      spec:
+        containers:
+        - image: quay.io/powercloud/docker-ce-build@sha256:20890c8770ff57f91eb9f16fbb9f9e979958cd02b2a3c03795d2587e7b58a9db
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:18]: Why not use `latest` or the proper version tag here instead of SHA tag.
I see this SHA mapped to `0.7-ppc64le` tag at https://quay.io/repository/powercloud/docker-ce-build?tab=tags.

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml [hunk 2]
```diff
@@ -0,0 +1,166 @@
+periodics:
+    - name: periodic-config-docker
+      cluster: k8s-ppc64le-cluster
+      cron: 0 0 * * *
+      decorate: true
+      decoration_config:
+        timeout: 2h
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:7]: The default 2h timeout is picked up from here https://github.com/ppc64le-cloud/test-infra/blob/master/config/prow/config.yaml#L74.
Needn't be mentioned in this file.

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          args:
+          - -c
+          - |
+            set -o errexit
+            set -o nounset
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-unit-test-docker *"
+            chmod ug+x $PWD/upstream-master-ci/prow-unit-test-docker.sh
+            $PWD/upstream-master-ci/prow-unit-test-docker.sh
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:122]: Hope junit xml file generation of results is taken care of in the script. Only then, we can expect a meaningful, easy to read test results in UI like in 
image

**Review by Rajalakshmi-Girish — CHANGES_REQUESTED** [review]
Also, make sure your shell scripts that run tests return non-zero code even on test case failures and not only for the script flow failure. (There were instances when such scripts returned zero exit code inspite of testcase failures)

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            set -o pipefail
+            set -o xtrace
+            echo "* Start prow-info-docker *"
+            chmod ug+x $PWD/upstream-master-ci/prow-info-docker.sh
+            $PWD/upstream-master-ci/prow-info-docker.sh
+            rc=$?
+            [ $rc != 0 ] && echo "ERROR: prow-info-docker exited with code:$rc"
+            exit $rc
+          securityContext:
+            privileged: true
```
Comment (Rajalakshmi-Girish) [inline config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:40]: Is the below required for all four jobs?
```
securityContext:
  privileged: true
```

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml (hunk 1 above)
Comment (AshwinHIBM) [reply config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:18]: We have made this move for all our jobs post a discussion where we concluded that hashes are immutable, hence more secure from intentional/accidental re-tagging.

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml (hunk 4 above)
Comment (AshwinHIBM) [reply config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:40]: Yes, `dockerd` refuses to start otherwise.
`failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: iptables failed: iptables -t nat -N DOCKER: iptables v1.8.9 (nf_tables): Could not fetch rule set generation id: Permission denied (you must be root)`

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml (hunk 3 above)
Comment (AshwinHIBM) [reply config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml:116]: We had a discussion about this and here's a summary:
We're running a large number of unit tests against github.com/moby/moby (3000+) with a negligible failure rate (a.t.m. a single test fails). Such a small number of failures are bound to happen and so we're really only concerned about the tests running and us being able to see the results somewhere, the UI logs contain all the information we need, we don't need the report. That said, we're going to work on getting it in the proper format in the near future. Also, just to add, the tests we're running are in [moby](github.com/moby/moby) and not in our repository anyway, we're just issuing a `make` command that runs these tests.

File: config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml (hunk 2 above)
Comment (AshwinHIBM) [reply config/jobs/ppc64le-cloud/build-docker/periodic-ci-docker.yaml]: Thanks for that, I have removed it from the job.

**Comment by AshwinHIBM** [comment]
> Also, make sure your shell scripts that run tests return non-zero code even on test case failures and not only for the script flow failure. (There were instances when such scripts returned zero exit code inspite of testcase failures)

This definitely makes sense to us, but for now our priority is to get these tests running first and then later try to patch this behavior. PTAL at https://github.com/ppc64le-cloud/test-infra/pull/447#discussion_r1497028238 where I've mentioned more details.

**Comment by Rajalakshmi-Girish** [comment]
@AshwinHIBM Can you please squash the multiple commits in this PR to one single commit?

**Comment by Rajalakshmi-Girish** [comment]
@AshwinHIBM I have initiated the `test-pj` run for all 4 jobs on production.
Will update here the links once they are complete

**Comment by AshwinHIBM** [comment]
> [periodic-unit-test-docker] - Are you fine not having Junit result tab? Also, the Run PASSED inspite of UT failures! Please check the script that returns exit code to job.

Like I mentioned earlier, the failures don't come as a surprise to us. We're only issuing a `make` command against [moby](https://github.com/moby/moby/) 
https://github.com/ppc64le-cloud/docker-ce-build/pull/182/files#diff-7a34ff4012a0fa233c41d7d2a14da8560e91e78206c9497215f888f18f923bafR2
If you look at the log you'll see there's not much we can do about them in our repository (at least at the moment). We just want the tests to run, that's all.
As far as the reports are concerned, that's a nice to have for us and I'm going to be working on that in a follow-up pull request because we need to have these tests running first.

> [periodic-integration-test-docker] - @AshwinHIBM Timedout after 2h run

I have adjusted the timeout to 3 hours, I'm not sure why it took that long. If it works with 3h we may need to think about reducing the frequency of this job maybe?
Thanks for the run and notifying us about the results.

**Comment by Rajalakshmi-Girish** [comment]
> If you look at the log you'll see there's not much we can do about them in our repository (at least at the moment). We just want the tests to run, that's all.

Yes, I understand that you are running the tests directly from upstream repo.
I have been insisting on having the Junit tab generated with the correct job status to make the job's existence meaningful and make it easy for you to analyze.

I see the moby repo already generating the junit xml files in https://github.com/moby/moby/blob/master/hack/test/unit.
These files seem to exist in the `bundles` directory. You may simply copy the bundles dir into ${ARTIFACTS} in the job yaml.
This would give you the correct job result with an easy summary of failures.

**Comment by Rajalakshmi-Girish** [comment]
> I have adjusted the timeout to 3 hours, I'm not sure why it took that long. If it works with 3h we may need to think about reducing the frequency of this job maybe?

I shall rerun the job and update here.

**Comment by Rajalakshmi-Girish** [comment]
> I see the moby repo already generating the junit xml files in https://github.com/moby/moby/blob/master/hack/test/unit.
> These files seem to exist in the `bundles` directory. You may simply copy the bundles dir into ${ARTIFACTS} in the job yaml.
> This would give you the correct job result with an easy summary of failures.

@AshwinHIBM Like the `cp` at the end of line https://github.com/ppc64le-cloud/test-infra/blob/master/config/jobs/periodic/containerd/test-containerd-periodics.yaml#L32

**Comment by Rajalakshmi-Girish** [comment]
https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-integration-test-docker/1760921594089181184 completed in 2h 33min.
Even this job PASSED though there are failures.

**Comment by Rajalakshmi-Girish** [comment]
@AshwinHIBM I have re-run the jobs after your recent script changes.

As discussed, I see the unit tests being displayed in the Junit tab,
and Integration suite returning failure status based on test results.

Please go through the below links and confirm if they appear as you expect them to.

[periodic-config-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-config-docker/1765571207446401024)
[periodic-build-dev-image-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-build-dev-image-docker/1765571717444407296)
[periodic-unit-test-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-unit-test-docker/1765574952821460992)
[periodic-integration-test-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-integration-test-docker/1765575219092656128)

**Comment by AshwinHIBM** [comment]
> @AshwinHIBM I have re-run the jobs after your recent script changes.
> 
> As discussed, I see the unit tests being displayed in the Junit tab, and Integration suite returning failure status based on test results.
> 
> Please go through the below links and confirm if they appear as you expect them to.
> 
> [periodic-config-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-config-docker/1765571207446401024) [periodic-build-dev-image-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-build-dev-image-docker/1765571717444407296) [periodic-unit-test-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-unit-test-docker/1765574952821460992) [periodic-integration-test-docker](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/logs/periodic-integration-test-docker/1765575219092656128)

These results are in line with our expectations. Thanks for the runs.

**Review by Rajalakshmi-Girish — APPROVED** [review]
/lgtm

**Comment by Rajalakshmi-Girish** [comment]
> @AshwinHIBM: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> pull-test-infra-yamllint 	[413b41e](https://github.com/ppc64le-cloud/test-infra/commit/413b41e9f6014ea561dbdf8f840aac3f58b7116b) 	[link](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/pr-logs/pull/ppc64le-cloud_test-infra/447/pull-test-infra-yamllint/1765811087560675328) 	true 	`/test pull-test-infra-yamllint`
> 
> [Full PR test history](https://prow.ppc64le-cloud.cis.ibm.net/pr-history?org=ppc64le-cloud&repo=test-infra&pr=447). [Your PR dashboard](https://prow.ppc64le-cloud.cis.ibm.net/pr?query=is:pr+state:open+author:AshwinHIBM).

@AshwinHIBM Looks like there are trailing spaces.
Please correct and re-submit.

**Comment by Rajalakshmi-Girish** [comment]
> @AshwinHIBM: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> pull-test-infra-yamllint 	[690d066](https://github.com/ppc64le-cloud/test-infra/commit/690d06634c42d5474f964919d4e8d12158db8a8e) 	[link](https://prow.ppc64le-cloud.cis.ibm.net/view/s3/ppc64le-prow-logs/pr-logs/pull/ppc64le-cloud_test-infra/447/pull-test-infra-yamllint/1765974259823284224) 	true 	`/test pull-test-infra-yamllint`
> 
> [Full PR test history](https://prow.ppc64le-cloud.cis.ibm.net/pr-history?org=ppc64le-cloud&repo=test-infra&pr=447). [Your PR dashboard](https://prow.ppc64le-cloud.cis.ibm.net/pr?query=is:pr+state:open+author:AshwinHIBM).

@AshwinHIBM It still fails. Can you please see the above link.
